sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -363,48 +364,120 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
+
+          case SaveMode.Append =>
+            // Check if the specified data source match the data source of the existing table.
+            val specifiedProvider = DataSource.lookupDataSource(source)
+            // TODO: Check that options from the resolved relation match the relation that we are
+            // inserting into (i.e. using the same compression).
+
+            // Pass a table identifier with database part, so that `lookupRelation` won't get temp
+            // views unexpectedly.
+            EliminateSubqueryAliases(catalog.lookupRelation(tableIdentWithDB)) match {
+              case l @ LogicalRelation(_: InsertableRelation | _: HadoopFsRelation, _, _) =>
+                // check if the file formats match
+                l.relation match {
+                  case r: HadoopFsRelation if r.fileFormat.getClass != specifiedProvider =>
+                    throw new AnalysisException(
+                      s"The file format of the existing table $tableIdent is " +
+                        s"`${r.fileFormat.getClass.getName}`. It doesn't match the specified " +
+                        s"format `$source`")
+                  case _ =>
+                }
+              case s: SimpleCatalogRelation if DDLUtils.isDatasourceTable(s.metadata) => // OK.
+              case c: CatalogRelation if c.catalogTable.provider == Some(DDLUtils.HIVE_PROVIDER) =>
+                throw new AnalysisException("Saving data in the Hive serde table " +
+                  s"${c.catalogTable.identifier} is not supported yet. Please use the " +
+                  "insertInto() API as an alternative..")
+              case o =>
+                throw new AnalysisException(s"Saving data in ${o.toString} is not supported.")
+            }
+
+            val existingSchema = tableMeta.schema
+            if (df.logicalPlan.schema.size != existingSchema.size) {
+              throw new AnalysisException(
+                s"The column number of the existing schema[$existingSchema] " +
+                  s"doesn't match the data schema[${df.logicalPlan.schema}]")
+            }
+
+            val specifiedPartCols = partitioningColumns.getOrElse(Nil)
+            val existingPartCols = tableMeta.partitionColumnNames
+            if (specifiedPartCols.map(_.toLowerCase) != existingPartCols.map(_.toLowerCase)) {
+              throw new AnalysisException("The partition columns of the existing table " +
+                s"$tableIdent are: [${existingPartCols.mkString(", ")}]. It doesn't match the " +
+                s"specified partition columns: [${specifiedPartCols.mkString(", ")}]")
+            }
+
+            // Because we are inserting into an existing table, we should respect the existing
+            // schema and adjust columns order of the given dataframe according to it.
+            df.select(existingSchema.map(f => Column(f.name)): _*)
+              .write.insertInto(tableIdentWithDB)
```

Comment:
Sorry, I made the mistake here. I deleted the comment after I realized it. 

`assertNotBucketed("insertInto")` is missing here. This is an existing bug, right?